File: sdk/ruby/lib/routeiq/handles.rb
require "digest"
require "json"
require "opentelemetry/sdk"

module RouteIQ
  COMPLETION_SUCCESS = "1"
  COMPLETION_FAILURE = "2"
  TOOL_SUCCESS = "1"
  TOOL_FAILURE = "2"

  PERMISSION = {
    "READ_ONLY"  => "1",
    "READ_WRITE" => "2",
    "PRIVILEGED" => "3"
  }.freeze

  # ── ToolHandle ──────────────────────────────────────────────────────────────

  class ToolHandle
    def initialize(step, name, args: {}, permission: "READ_ONLY")
      @step       = step
      @name       = name
      @args       = args
      @permission = permission
      @start      = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      @done       = false

      args_hash = Digest::SHA256.hexdigest(JSON.generate(args.sort.to_h))[0, 16]
      perm      = PERMISSION.fetch(permission, "1")
      riq       = step.task.riq

      step.task.record_tool(name)

      @span = riq.tracer.start_span("tool:#{name}", attributes: {
        "routeiq.event.type"             => "7",
        **riq.envelope(step.task, step),
        "routeiq.tool.name"              => name,
        "routeiq.tool.arguments_hash"    => args_hash,
        "routeiq.tool.permission_level"  => perm
      })
    end

    def success(latency_ms: nil, tokens_in: nil, tokens_out: nil)
      finish(TOOL_SUCCESS, latency_ms: latency_ms, tokens_in: tokens_in, tokens_out: tokens_out)
    end

    def fail(error_code: "", latency_ms: nil, retry_count: nil, tokens_in: nil, tokens_out: nil)
      finish(TOOL_FAILURE, error_code: error_code, latency_ms: latency_ms,
             retry_count: retry_count, tokens_in: tokens_in, tokens_out: tokens_out)
    end

    def end_span
      @span.finish unless @span.nil?
    end

    private

    def finish(status, error_code: "", latency_ms: nil, retry_count: nil,
               tokens_in: nil, tokens_out: nil)
      return if @done
      @done = true
      elapsed = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - @start) * 1000
      attrs = {
        "routeiq.tool.result_status" => status,
        "routeiq.tool.latency_ms"    => latency_ms || elapsed
      }
      attrs["routeiq.tool.error_code"]  = error_code  unless error_code.to_s.empty?
      attrs["routeiq.tool.retry_count"] = retry_count  if retry_count
      attrs["routeiq.tool.tokens_in"]   = tokens_in    if tokens_in
      attrs["routeiq.tool.tokens_out"]  = tokens_out   if tokens_out
      @span&.add_attributes(attrs)
    end
  end

  # ── StepHandle ──────────────────────────────────────────────────────────────

  class StepHandle
    attr_reader :step_id, :task

    def initialize(task, action: nil, rationale: nil, model: nil, index: 1)
      @task     = task
      @step_id  = SecureRandom.uuid
      @done     = false
      riq       = task.riq

      attrs = {
        "routeiq.event.type"  => "4",
        **riq.envelope(task, self),
        "routeiq.step.index"  => index
      }
      attrs["routeiq.step.selected_action"]   = action    if action
      attrs["routeiq.step.action_rationale"]  = rationale if rationale
      attrs["routeiq.step.model"]             = model     if model

      @span = riq.tracer.start_span("step:#{@step_id}", attributes: attrs)
    end

    def tool(name, args: {}, permission: "READ_ONLY", &block)
      handle = ToolHandle.new(self, name, args: args, permission: permission)
      if block_given?
        begin
          block.call(handle)
          handle.success unless handle.instance_variable_get(:@done)
        rescue => e
          handle.fail unless handle.instance_variable_get(:@done)
          raise
        ensure
          handle.end_span
        end
      else
        handle
      end
    end

    def guardrail(type, blocked)
      riq = @task.riq
      span = riq.tracer.start_span("guardrail:#{type}", attributes: {
        "routeiq.event.type"       => "9",
        **riq.envelope(@task, self),
        "routeiq.guardrail.type"    => type,
        "routeiq.guardrail.blocked" => blocked.to_s
      })
      span.finish
    end

    def replan(reason)
      @span&.add_attributes(
        "routeiq.replan.triggered" => "true",
        "routeiq.replan.reason"    => reason[0, 256]
      )
    end

    def complete(tokens_in: nil, tokens_out: nil)
      finish(COMPLETION_SUCCESS, tokens_in: tokens_in, tokens_out: tokens_out)
    end

    def fail(category: "")
      finish(COMPLETION_FAILURE, failure_category: category)
    end

    def end_span
      @span.finish unless @span.nil?
    end

    private

    def finish(status, failure_category: "", tokens_in: nil, tokens_out: nil)
      return if @done
      @done = true
      attrs = {"routeiq.step.completion_status" => status}
      attrs["routeiq.step.failure_category"] = failure_category unless failure_category.to_s.empty?
      attrs["routeiq.step.tokens_in"]        = tokens_in        if tokens_in
      attrs["routeiq.step.tokens_out"]       = tokens_out       if tokens_out
      @span&.add_attributes(attrs)
    end
  end

  # ── TaskHandle ──────────────────────────────────────────────────────────────

  class TaskHandle
    attr_reader :task_id, :run_id, :riq

    def initialize(riq, intent, task_type: nil)
      @riq          = riq
      @intent       = intent
      @task_type    = task_type
      @task_id      = SecureRandom.uuid
      @run_id       = SecureRandom.uuid
      @done         = false
      @step_index   = 0
      @tool_sequence = []

      attrs = {
        "routeiq.event.type"        => "1",
        **riq.envelope(self),
        "routeiq.task.input_intent" => intent[0, 256]
      }
      attrs["routeiq.task.type"] = task_type if task_type

      @span = riq.tracer.start_span("task:#{@task_id}", attributes: attrs)
    end

    def record_tool(name)
      @tool_sequence << name
    end

    def max_same_tool_count
      return 0 if @tool_sequence.empty?
      max_count = cur = 1
      (1...@tool_sequence.length).each do |i|
        cur = @tool_sequence[i] == @tool_sequence[i - 1] ? cur + 1 : 1
        max_count = cur if cur > max_count
      end
      max_count
    end

    def step(action: nil, rationale: nil, model: nil, &block)
      @step_index += 1
      handle = StepHandle.new(self, action: action, rationale: rationale,
                              model: model, index: @step_index)
      if block_given?
        begin
          block.call(handle)
          handle.complete unless handle.instance_variable_get(:@done)
        rescue => e
          handle.fail unless handle.instance_variable_get(:@done)
          raise
        ensure
          handle.end_span
        end
      else
        handle
      end
    end

    def escalate(reason: nil, target: nil)
      riq = @riq
      span = riq.tracer.start_span("escalation:#{@task_id}", attributes: {
        "routeiq.event.type"             => "8",
        **riq.envelope(self),
        "routeiq.escalation.triggered"   => "true"
      }.tap do |attrs|
        attrs["routeiq.escalation.reason"] = reason[0, 256] if reason
        attrs["routeiq.escalation.target"] = target          if target
      end)
      span.finish
    end

    def complete(tokens: 0, tokens_in: nil, tokens_out: nil, cost_usd: nil, cohort: nil)
      finish(COMPLETION_SUCCESS, tokens: tokens, tokens_in: tokens_in,
             tokens_out: tokens_out, cost_usd: cost_usd, cohort: cohort)
    end

    def fail(category: "")
      finish(COMPLETION_FAILURE, failure_category: category)
    end

    def end_span
      @span.finish unless @span.nil?
    end

    private

    def finish(status, tokens: 0, tokens_in: nil, tokens_out: nil,
               cost_usd: nil, cohort: nil, failure_category: "")
      return if @done
      @done = true
      attrs = {"routeiq.task.completion_status" => status}
      attrs["routeiq.task.tokens_in"]      = tokens_in   if tokens_in
      attrs["routeiq.task.tokens_out"]     = tokens_out  if tokens_out
      total = tokens > 0 ? tokens : ((tokens_in || 0) + (tokens_out || 0))
      attrs["routeiq.task.total_tokens"]   = total       if total > 0
      attrs["routeiq.task.cost_usd"]       = cost_usd    if cost_usd
      attrs["routeiq.task.cohort"]         = cohort      if cohort
      attrs["routeiq.task.failure_category"] = failure_category unless failure_category.to_s.empty?
      same = max_same_tool_count
      attrs["routeiq.same_tool_count"] = same if same > 1
      @span&.add_attributes(attrs)
    end
  end
end

File: sdk/python/src/routeiq/handles.py
"""
TaskHandle, StepHandle, ToolHandle — sync context managers for the RouteIQ SDK.

Attribute keys match conventions/telemetry.yaml in routeiq-schema.
Enum values match entities.proto (SUCCESS=1, FAILURE=2, TOOL_SUCCESS=1, TOOL_FAILURE=2).
"""

import hashlib
import json
import time
import uuid
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from .client import RouteIQ

# ── Enum values (mirror entities.proto) ──────────────────────────────────────

_COMPLETION_SUCCESS = "1"
_COMPLETION_FAILURE = "2"
_TOOL_SUCCESS = "1"
_TOOL_FAILURE = "2"

PERMISSION = {
    "READ_ONLY":  "1",
    "READ_WRITE": "2",
    "PRIVILEGED": "3",
}


# ── ToolHandle ────────────────────────────────────────────────────────────────

class ToolHandle:
    """Context manager for a single tool invocation inside a step."""

    def __init__(
        self,
        step: "StepHandle",
        name: str,
        args: Optional[dict] = None,
        permission: str = "READ_ONLY",
    ):
        self._step = step
        self.name = name
        self._args = args or {}
        self._permission = PERMISSION.get(permission, "1")
        self._start: float = 0.0
        self._span_cm = None
        self._span = None
        self._done = False

    def __enter__(self) -> "ToolHandle":
        riq = self._step._task._riq
        self._start = time.monotonic()
        args_hash = hashlib.sha256(
            json.dumps(self._args, sort_keys=True).encode()
        ).hexdigest()[:16]
        self._span_cm = riq._tracer.start_as_current_span(f"tool:{self.name}")
        self._span = self._span_cm.__enter__()
        self._span.set_attributes({
            "routeiq.event.type": "7",  # TOOL_CALLED
            **riq._envelope(self._step._task, self._step),
            "routeiq.tool.name": self.name,
            "routeiq.tool.arguments_hash": args_hash,
            "routeiq.tool.permission_level": self._permission,
        })
        # Register with task for same_tool_count tracking
        self._step._task._record_tool(self.name)
        return self

    def success(
        self,
        latency_ms: Optional[float] = None,
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
    ) -> None:
        """Record a successful tool result."""
        self._finish(_TOOL_SUCCESS, latency_ms=latency_ms, tokens_in=tokens_in, tokens_out=tokens_out)

    def fail(
        self,
        error_code: str = "",
        latency_ms: Optional[float] = None,
        retry_count: Optional[int] = None,
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
    ) -> None:
        """Record a failed tool result."""
        self._finish(_TOOL_FAILURE, error_code=error_code, latency_ms=latency_ms,
                     retry_count=retry_count, tokens_in=tokens_in, tokens_out=tokens_out)

    def _finish(
        self,
        status: str,
        error_code: str = "",
        latency_ms: Optional[float] = None,
        retry_count: Optional[int] = None,
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
    ):
        if self._done:
            return
        self._done = True
        elapsed = (time.monotonic() - self._start) * 1000
        attrs: dict = {
            "routeiq.tool.result_status": status,
            "routeiq.tool.latency_ms": latency_ms if latency_ms is not None else elapsed,
        }
        if error_code:                   attrs["routeiq.tool.error_code"]  = error_code
        if retry_count is not None:      attrs["routeiq.tool.retry_count"] = retry_count
        if tokens_in is not None:        attrs["routeiq.tool.tokens_in"]   = tokens_in
        if tokens_out is not None:       attrs["routeiq.tool.tokens_out"]  = tokens_out
        if self._span:
            self._span.set_attributes(attrs)

    def __exit__(self, exc_type, exc_val, exc_tb):
        if not self._done:
            elapsed = (time.monotonic() - self._start) * 1000
            if exc_type is not None:
                self.fail(latency_ms=elapsed)
            else:
                self.success(latency_ms=elapsed)
        self._span_cm.__exit__(exc_type, exc_val, exc_tb)
        return False


# ── StepHandle ────────────────────────────────────────────────────────────────

class StepHandle:
    """Context manager for one reasoning step within a task."""

    def __init__(
        self,
        task: "TaskHandle",
        action: Optional[str] = None,
        rationale: Optional[str] = None,
        index: int = 1,
        model: Optional[str] = None,
    ):
        self._task = task
        self.step_id = str(uuid.uuid4())
        self._action = action
        self._rationale = rationale
        self._index = index
        self._model = model
        self._span_cm = None
        self._span = None
        self._done = False

    def __enter__(self) -> "StepHandle":
        riq = self._task._riq
        self._span_cm = riq._tracer.start_as_current_span(f"step:{self.step_id}")
        self._span = self._span_cm.__enter__()
        attrs = {
            "routeiq.event.type": "4",  # STEP_STARTED
            **riq._envelope(self._task, self),
        }
        if self._action:    attrs["routeiq.step.selected_action"]  = self._action
        if self._rationale: attrs["routeiq.step.action_rationale"] = self._rationale
        if self._model:     attrs["routeiq.step.model"]            = self._model
        attrs["routeiq.step.index"] = self._index
        self._span.set_attributes(attrs)
        return self

    def tool(
        self,
        name: str,
        args: Optional[dict] = None,
        permission: str = "READ_ONLY",
    ) -> ToolHandle:
        """Start a tool call within this step."""
        return ToolHandle(self, name=name, args=args, permission=permission)

    def guardrail(self, type: str, blocked: bool) -> None:
        """Emit a guardrail check span. Call whenever a policy/guardrail fires."""
        riq = self._task._riq
        with riq._tracer.start_as_current_span(f"guardrail:{type}") as span:
            span.set_attributes({
                "routeiq.event.type": "9",
                **riq._envelope(self._task, self),
                "routeiq.guardrail.type":    type,
                "routeiq.guardrail.blocked": str(blocked).lower(),
            })

    def replan(self, reason: str) -> None:
        """Mark that the agent replanned mid-step."""
        if self._span:
            self._span.set_attributes({
                "routeiq.replan.triggered": "true",
                "routeiq.replan.reason":    reason[:256],
            })

    def complete(self, tokens_in: Optional[int] = None, tokens_out: Optional[int] = None) -> None:
        """Mark step as successfully completed."""
        self._finish(_COMPLETION_SUCCESS, tokens_in=tokens_in, tokens_out=tokens_out)

    def fail(self, category: str = "") -> None:
        """Mark step as failed."""
        self._finish(_COMPLETION_FAILURE, failure_category=category)

    def _finish(
        self,
        status: str,
        failure_category: str = "",
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
    ):
        if self._done:
            return
        self._done = True
        attrs: dict = {"routeiq.step.completion_status": status}
        if failure_category:        attrs["routeiq.step.failure_category"] = failure_category
        if tokens_in is not None:   attrs["routeiq.step.tokens_in"]        = tokens_in
        if tokens_out is not None:  attrs["routeiq.step.tokens_out"]       = tokens_out
        if self._span:
            self._span.set_attributes(attrs)

    def __exit__(self, exc_type, exc_val, exc_tb):
        if not self._done:
            if exc_type is not None:
                self.fail()
            else:
                self.complete()
        self._span_cm.__exit__(exc_type, exc_val, exc_tb)
        return False


# ── TaskHandle ────────────────────────────────────────────────────────────────

class TaskHandle:
    """Context manager for a complete agent task."""

    def __init__(
        self,
        riq: "RouteIQ",
        intent: str,
        task_type: Optional[str] = None,
    ):
        self._riq = riq
        self.intent = intent
        self.task_type = task_type
        self.task_id = str(uuid.uuid4())
        self.run_id = str(uuid.uuid4())
        self._span_cm = None
        self._span = None
        self._done = False
        self._step_index = 0
        self._tool_sequence: list = []

    def __enter__(self) -> "TaskHandle":
        self._span_cm = self._riq._tracer.start_as_current_span(f"task:{self.task_id}")
        self._span = self._span_cm.__enter__()
        attrs = {
            "routeiq.event.type": "1",  # TASK_STARTED
            **self._riq._envelope(self),
            "routeiq.task.input_intent": self.intent[:256],
        }
        if self.task_type:
            attrs["routeiq.task.type"] = self.task_type
        self._span.set_attributes(attrs)
        return self

    def step(
        self,
        action: Optional[str] = None,
        rationale: Optional[str] = None,
        model: Optional[str] = None,
    ) -> StepHandle:
        """Start a reasoning step within this task."""
        self._step_index += 1
        return StepHandle(self, action=action, rationale=rationale,
                          index=self._step_index, model=model)

    def escalate(self, reason: Optional[str] = None, target: Optional[str] = None) -> None:
        """Emit a human-escalation span. Call when handing off to a human operator."""
        riq = self._riq
        with riq._tracer.start_as_current_span(f"escalation:{self.task_id}") as span:
            attrs: dict = {
                "routeiq.event.type": "8",
                **riq._envelope(self),
                "routeiq.escalation.triggered": "true",
            }
            if reason: attrs["routeiq.escalation.reason"] = reason[:256]
            if target: attrs["routeiq.escalation.target"] = target
            span.set_attributes(attrs)

    def _record_tool(self, name: str) -> None:
        """Called by ToolHandle to track tool call sequence for same_tool_count."""
        self._tool_sequence.append(name)

    def _max_same_tool_count(self) -> int:
        if not self._tool_sequence:
            return 0
        max_count = cur = 1
        for i in range(1, len(self._tool_sequence)):
            cur = cur + 1 if self._tool_sequence[i] == self._tool_sequence[i - 1] else 1
            if cur > max_count:
                max_count = cur
        return max_count

    def complete(
        self,
        tokens: int = 0,
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
        cost_usd: Optional[float] = None,
        cohort: Optional[str] = None,
    ) -> None:
        """Mark task as successfully completed."""
        self._finish(_COMPLETION_SUCCESS, tokens=tokens, tokens_in=tokens_in,
                     tokens_out=tokens_out, cost_usd=cost_usd, cohort=cohort)

    def fail(self, category: str = "") -> None:
        """Mark task as failed."""
        self._finish(_COMPLETION_FAILURE, failure_category=category)

    def _finish(
        self,
        status: str,
        tokens: int = 0,
        tokens_in: Optional[int] = None,
        tokens_out: Optional[int] = None,
        cost_usd: Optional[float] = None,
        cohort: Optional[str] = None,
        failure_category: str = "",
    ):
        if self._done:
            return
        self._done = True
        attrs: dict = {"routeiq.task.completion_status": status}
        if tokens_in is not None:   attrs["routeiq.task.tokens_in"]  = tokens_in
        if tokens_out is not None:  attrs["routeiq.task.tokens_out"] = tokens_out
        total = tokens or (
            (tokens_in + tokens_out) if tokens_in is not None and tokens_out is not None else 0
        )
        if total:                   attrs["routeiq.task.total_tokens"]    = total
        if cost_usd is not None:    attrs["routeiq.task.cost_usd"]        = cost_usd
        if cohort:                  attrs["routeiq.task.cohort"]          = cohort
        if failure_category:        attrs["routeiq.task.failure_category"] = failure_category
        # Emit max consecutive same-tool count for loop detection
        same_tool_count = self._max_same_tool_count()
        if same_tool_count > 1:     attrs["routeiq.same_tool_count"] = same_tool_count
        if self._span:
            self._span.set_attributes(attrs)

    def __exit__(self, exc_type, exc_val, exc_tb):
        if not self._done:
            if exc_type is not None:
                self.fail()
            else:
                self.complete()
        self._span_cm.__exit__(exc_type, exc_val, exc_tb)
        return False
